clang/include/clang/Tooling/DependencyScanning/DependencyScanningCASFilesystem.h

```diff
@@ -72,11 +75,17 @@ class DependencyScanningCASFilesystem : public llvm::cas::ThreadSafeFileSystem {
   Optional<ArrayRef<dependency_directives_scan::Directive>>
   getDirectiveTokens(const Twine &Path);
 
+  /// Set a \c DiagnosticsEngine to use if there are errors scanning for
+  /// directives. If null, only filesystem errors will be reported via
+  /// the usual \c error_code returns.
+  void setDiagnostics(DiagnosticsEngine *Diags) { this->Diags = Diags; }
```

Comment:
I don't love this `setDiagnostics` code, but `DependencyScanningCASFilesystem` is shared across multiple scanning TUs whereas the diagnostics are scoped to a single TU.  We could maybe extract the caching layer from the FS to share and create a new FS for each TU so the diagnostics are available at construction time, but that seems like a big refactoring and not really worth it just for this case.